# Introduction to Numpy

NumPy is THE fundamental library for scientific calculation in python. You can play with this library to do deeplearning but NumPy is not the best choice... Nevertheless, most scientific libs rely on NumPy conventions and APIs so it is important to have some knowledges about it. If you are familiar with NumPy, you can skip this section and go to the section _Introduction to Tensorflow_

## The `ndarray` class

To use NumPy you should import it with the following command

In [1]:
import numpy as np

Now you can use Numpy with the shortcut `np`.

The fundamental class of NumPy is `ndarray`. It represents table of items, with the following constraints:

* It's _multidimensional_ (1d, 2d, 3d, ..., nd),
* It's _homogeneous_, that is, all items inside the table should belong to the same type.



In [2]:
# Ndarray instanciation from known values
a = np.array([[1., 2., 3.], [3., 4., .5]])
a

array([[1. , 2. , 3. ],
       [3. , 4. , 0.5]])

In [3]:
# Type of a
type(a)

numpy.ndarray

In [4]:
# 'Rank' as mention in NumPy doc or number of dimensions
a.ndim

2

In [5]:
# Shape of the ndarray
a.shape

(2, 3)

In [6]:
# Total number of items
a.size

6

In [7]:
# Item type
a.dtype

dtype('float64')

In [8]:
# Actual data of the table
a.data

## Creation of a `ndarray`

The basic constructors of `ndarray`are :

* `numpy.array(object, dtype=None, copy=True, order=’K’, subok=False, ndmin=0)`
Create an array from known values
* `numpy.zeros(shape, dtype=float, order=’C)`
Create an array full of zeros
* `numpy.ones(shape, dtype=None, order=’C’)`
Create an array full of ones

`dtype` determines the type of each element, `order` indicates how elements are organized into `data` .

**Take Care !** `dtype` is determined at instanciation and can not be changed after.


In [9]:
np.array([[1., 2., 3.], [3., 4., .5]])

array([[1. , 2. , 3. ],
       [3. , 4. , 0.5]])

In [10]:
np.zeros((5, 3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [11]:
np.ones((2, 2, 3), dtype='int')

array([[[1, 1, 1],
        [1, 1, 1]],

       [[1, 1, 1],
        [1, 1, 1]]])

Some other useful methods are :

* `numpy.arange([start, ]stop, [step, ]dtype=None)`
* `numpy.linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None)`
* `numpy.logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None)`
* `numpy.eye(N, M=None, k=0, dtype=float)`
* `numpy.random.randn(d0, d1, ..., dn)`

More creation routines are available [here](https://docs.scipy.org/doc/numpy/reference/routines.array-creation.html).

In [12]:
np.arange(5, 10, 1)

array([5, 6, 7, 8, 9])

In [13]:
np.arange(0, 10, 2)

array([0, 2, 4, 6, 8])

In [14]:
np.linspace(0, 10, 20)

array([ 0.        ,  0.52631579,  1.05263158,  1.57894737,  2.10526316,
        2.63157895,  3.15789474,  3.68421053,  4.21052632,  4.73684211,
        5.26315789,  5.78947368,  6.31578947,  6.84210526,  7.36842105,
        7.89473684,  8.42105263,  8.94736842,  9.47368421, 10.        ])

In [15]:
np.logspace(1, 10, 20)

array([1.00000000e+01, 2.97635144e+01, 8.85866790e+01, 2.63665090e+02,
       7.84759970e+02, 2.33572147e+03, 6.95192796e+03, 2.06913808e+04,
       6.15848211e+04, 1.83298071e+05, 5.45559478e+05, 1.62377674e+06,
       4.83293024e+06, 1.43844989e+07, 4.28133240e+07, 1.27427499e+08,
       3.79269019e+08, 1.12883789e+09, 3.35981829e+09, 1.00000000e+10])

In [16]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [17]:
np.eye(3, 4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.]])

In [18]:
np.random.randn(3, 4)
# !!! shape is given dimension by dimension as arguments not in one tuple

array([[ 0.60616019,  1.08343872, -0.2240689 ,  0.37609485],
       [ 1.17988919, -1.56132697, -0.78329256, -0.89763815],
       [-0.42291922,  0.94128562, -1.12592912, -0.623465  ]])

## Indexation / Slicing

### Monodimensional indexation

Indexing and slicing are done with the operator `[]` as for list.

In [19]:
a = np.random.randn(10)
a

array([ 1.16433506, -0.32221425,  0.25778607,  0.48871717, -0.23922105,
        1.20660032, -0.55452082, -0.27195536, -0.10656522,  1.64121768])

In [20]:
# First item
a[0]

1.164335058796136

In [21]:
# Last item
a[-1]

1.641217681220214

In [22]:
# From item 2 to item 5 (excluded !)
a[2:5]

array([ 0.25778607,  0.48871717, -0.23922105])

In [23]:
# Eliptic formulation
# 3 first items
a[:3]

array([ 1.16433506, -0.32221425,  0.25778607])

In [24]:
# Starting from the 4th item
a[3:]

array([ 0.48871717, -0.23922105,  1.20660032, -0.55452082, -0.27195536,
       -0.10656522,  1.64121768])

In [25]:
# All items
a[:]

array([ 1.16433506, -0.32221425,  0.25778607,  0.48871717, -0.23922105,
        1.20660032, -0.55452082, -0.27195536, -0.10656522,  1.64121768])

In [26]:
# With a step
a[2:8:2]

array([ 0.25778607, -0.23922105, -0.55452082])

In [27]:
# Reverse
a[::-1]

array([ 1.64121768, -0.10656522, -0.27195536, -0.55452082,  1.20660032,
       -0.23922105,  0.48871717,  0.25778607, -0.32221425,  1.16433506])

### Multidimensional indexation

In [28]:
b = np.random.randn(3, 4, 5)
b

array([[[-0.57764691,  0.0725287 ,  0.17787888,  1.02463507,
         -1.03600176],
        [ 0.3939148 ,  0.31163917, -1.14274504,  0.32648244,
         -1.64760117],
        [-1.23914186, -0.4113622 ,  0.55471875,  1.6606401 ,
         -1.03489785],
        [ 1.31944944, -0.06558464, -0.57498817, -0.52597785,
          0.95018486]],

       [[-0.94177535, -0.65571898, -0.24009783,  0.11925457,
          2.48079609],
        [ 0.11259129,  1.19217211,  1.5873622 , -0.19737594,
          1.63659061],
        [ 0.95456193,  1.12270818, -1.13270826,  0.93410428,
         -0.74040062],
        [ 1.63040502,  0.3596341 , -0.98242825,  0.03992308,
         -1.16401228]],

       [[ 0.55343368, -0.43444025, -1.2324443 , -2.00674008,
         -0.48282975],
        [-0.35611856,  0.50916111,  0.71659451,  0.58433746,
          1.14835663],
        [ 0.30524929,  0.50597895, -1.31456629,  0.41157335,
          0.31026184],
        [-0.21810683, -0.87837479,  2.8229405 ,  1.17932119,
         -1

In [29]:
# First item on each axis
b[0, 0, 0]

-0.5776469105994826

In [30]:
# With an interval and ann elipse
b[:, 1, 2:5]

array([[-1.14274504,  0.32648244, -1.64760117],
       [ 1.5873622 , -0.19737594,  1.63659061],
       [ 0.71659451,  0.58433746,  1.14835663]])

In [31]:
# a[2] is equivalent to a[2,:,:]
b[2]

array([[ 0.55343368, -0.43444025, -1.2324443 , -2.00674008, -0.48282975],
       [-0.35611856,  0.50916111,  0.71659451,  0.58433746,  1.14835663],
       [ 0.30524929,  0.50597895, -1.31456629,  0.41157335,  0.31026184],
       [-0.21810683, -0.87837479,  2.8229405 ,  1.17932119, -1.0779881 ]])

In [32]:
# Multiple elipses : c[1,...,2] is equivalent to c[1,:,:,2] on 4-D array
c = np.random.randn(2, 2, 2, 3)
c

array([[[[ 0.59505743, -0.15143831,  0.94895997],
         [-0.53794754,  0.71656876, -0.8497542 ]],

        [[ 0.91541166, -0.28991024, -0.45462799],
         [-1.23738945,  1.44107894,  0.43814181]]],


       [[[-0.61208687,  0.13369129, -0.34339933],
         [-0.52188227,  0.69125424, -1.24580836]],

        [[ 2.0771369 ,  1.08880695,  0.4691944 ],
         [ 0.15204394, -0.78349188,  0.0889454 ]]]])

In [33]:
c[1, ..., 2]

array([[-0.34339933, -1.24580836],
       [ 0.4691944 ,  0.0889454 ]])

In [34]:
c[1, :, :, 2]

array([[-0.34339933, -1.24580836],
       [ 0.4691944 ,  0.0889454 ]])

 ### Indexation implies cut in dimension ! (Warning for Matlab users)
 
 Important for matrix operation (multiplication...)

In [35]:
a = np.random.randn(4, 3)
a

array([[ 0.24711917, -0.83832244,  0.82718966],
       [ 0.49585881, -0.56881472,  0.97036443],
       [-2.24873211,  0.17617817, -0.97267131],
       [ 0.56953724, -0.83545757, -0.74184107]])

In [36]:
b = a[:, 0]
b

array([ 0.24711917,  0.49585881, -2.24873211,  0.56953724])

In [37]:
# b has shape (4,) not (4,1)
b.shape

(4,)

In [38]:
c = a[0, :]
c

array([ 0.24711917, -0.83832244,  0.82718966])

In [39]:
# c has shape (3,) not (1,3)
c.shape

(3,)

In [40]:
# Meanwhile using slice and not index preserves dimension
d = a[0:1, :]
d

array([[ 0.24711917, -0.83832244,  0.82718966]])

In [41]:
d.shape

(1, 3)

## Assignation

Assignation is performed by the operator `=`. Item or a sub-array can be targeted.


In [42]:
a = np.array([[1, 2, 3], [4, 5, 6]], dtype=int)
a

array([[1, 2, 3],
       [4, 5, 6]])

In [43]:
a[0, 0] = 10
a

array([[10,  2,  3],
       [ 4,  5,  6]])

In [44]:
a[0:2, 1:3] = np.ones((2, 2))
a

array([[10,  1,  1],
       [ 4,  1,  1]])

**Take Care !** `dtype` is determined at instanciation and can not be changed after.

In [45]:
# 1.75 will be downcast before assignation
a[1, 0] = 1.75
a

array([[10,  1,  1],
       [ 1,  1,  1]])

## Resize operation

Arrays can be reshaped by the `resize` method. That's an in-place operation:


In [46]:
a.resize((3, 2))
a

array([[10,  1],
       [ 1,  1],
       [ 1,  1]])

## References, view and copy


If `a` and `b` reference the same `ndarray`, all operation on `a`also applied to `b`. They share both data and metadata. 

If `c` is a view of `a`, they share the same data but not the metadata. For example shapes can be modified separately. But if we change the first element of `c`, the first element of `a` is also changed.

If `d` is a copy of `a`, all data and metadata are separated.


In [47]:
a = np.random.randn(4, 3)
a

array([[-2.2985055 , -1.06233181, -0.07975649],
       [-0.72489904, -0.66519295,  1.50343383],
       [-1.403098  , -0.18364155, -0.51704002],
       [ 0.47854268,  1.65404307, -0.32911513]])

In [48]:
# b is a reference to a
b = a
b[0, 0] = 1
a

array([[ 1.        , -1.06233181, -0.07975649],
       [-0.72489904, -0.66519295,  1.50343383],
       [-1.403098  , -0.18364155, -0.51704002],
       [ 0.47854268,  1.65404307, -0.32911513]])

In [49]:
# c is a view of a
c = a.view()
c.resize(3, 4)
c

array([[ 1.        , -1.06233181, -0.07975649, -0.72489904],
       [-0.66519295,  1.50343383, -1.403098  , -0.18364155],
       [-0.51704002,  0.47854268,  1.65404307, -0.32911513]])

In [50]:
# Shape of a is not affected
a

array([[ 1.        , -1.06233181, -0.07975649],
       [-0.72489904, -0.66519295,  1.50343383],
       [-1.403098  , -0.18364155, -0.51704002],
       [ 0.47854268,  1.65404307, -0.32911513]])

In [51]:
# But if we modify the last element of c, the last element of a is changed
c[2, 3] = 0
a

array([[ 1.        , -1.06233181, -0.07975649],
       [-0.72489904, -0.66519295,  1.50343383],
       [-1.403098  , -0.18364155, -0.51704002],
       [ 0.47854268,  1.65404307,  0.        ]])

In [52]:
# d is a copy of a
d = a.copy()
d

array([[ 1.        , -1.06233181, -0.07975649],
       [-0.72489904, -0.66519295,  1.50343383],
       [-1.403098  , -0.18364155, -0.51704002],
       [ 0.47854268,  1.65404307,  0.        ]])

In [53]:
d[0, 0] = 3
d

array([[ 3.        , -1.06233181, -0.07975649],
       [-0.72489904, -0.66519295,  1.50343383],
       [-1.403098  , -0.18364155, -0.51704002],
       [ 0.47854268,  1.65404307,  0.        ]])

In [54]:
# a was not modified by the assigniation on d
a

array([[ 1.        , -1.06233181, -0.07975649],
       [-0.72489904, -0.66519295,  1.50343383],
       [-1.403098  , -0.18364155, -0.51704002],
       [ 0.47854268,  1.65404307,  0.        ]])

## Shape manipulation

* `ndarray.resize(new shape, refcheck=True)`
Resize in-place
* `ndarray.reshape(shape, order=’C’)`
Return a view with a new shape
* `ndarray.ravel(order=’C’)`
Return a flatten view
* `ndarray.flatten(order=’C’)`
Return a flatten copy
* `numpy.concatenate((a1, a2, ...), axis=0)`
Return a concatenation of arrays along an existing axis
* `numpy.stack((a1, a2, ...), axis=0)`
Return a stack of arrays along a new axis



## Operations

Simple operations +, -, \*, \*\*, / operate item by item

In [55]:
a = np.random.randn(4, 3)
a

array([[ 0.11841939, -0.47439418,  1.30423905],
       [-0.88345197,  0.62403659,  2.01717748],
       [-1.18769552, -2.3955826 , -1.24063451],
       [ 0.15909517, -1.63103419, -1.4780723 ]])

In [56]:
b = np.random.randn(4, 3)
b

array([[ 0.14366041,  0.38297148, -0.98032245],
       [-0.2361172 , -0.47392511,  0.91582155],
       [ 1.51206043, -2.68432059, -2.03759569],
       [ 0.2726835 ,  0.66790811, -0.79897169]])

In [57]:
a + b

array([[ 0.26207981, -0.09142269,  0.32391659],
       [-1.11956917,  0.15011148,  2.93299904],
       [ 0.32436491, -5.07990319, -3.27823019],
       [ 0.43177868, -0.96312608, -2.27704399]])

In [58]:
a * b

array([[ 0.01701218, -0.18167944, -1.27857482],
       [ 0.2085982 , -0.29574661,  1.84737461],
       [-1.79586739,  6.4305117 ,  2.52791152],
       [ 0.04338263, -1.08938097,  1.18093793]])

In [59]:
a ** 2

array([[0.01402315, 0.22504983, 1.70103949],
       [0.78048738, 0.38942166, 4.06900501],
       [1.41062065, 5.73881601, 1.53917398],
       [0.02531127, 2.66027253, 2.18469772]])

Arrays can be viewed as set where we can get min/max of all items.

In [60]:
a.min()

-2.3955826025517113

In [61]:
a.max()

2.0171774848388875

In [62]:
# Position of the min in the flatten view of the array
a.argmin()

7

In [63]:
a.argmax()

5

If you want to compute an extremum along a particular axis, you should precise  `axis` in argument. As indexing, this reduce the dimension of the array. If you want to keep the same number of dimension, you should set the `keepdims` argument to `True`.

In [64]:
a.min(axis=1)

array([-0.47439418, -0.88345197, -2.3955826 , -1.63103419])

In [65]:
a.min(axis=1, keepdims=True)

array([[-0.47439418],
       [-0.88345197],
       [-2.3955826 ],
       [-1.63103419]])

## Broadcasting

Broadcasting is a mechanism to automatically tile arrays of incompatible dimensions before an operation.
This is a powerfull mechanism (you don't have to use `repmat` as in Matlab) but it can hide dimensionality errors.

In [66]:
a = np.array([[1, 2, 3, 4]])
a.shape

(1, 4)

In [67]:
b = np.array([[5], [6], [7]])
b.shape

(3, 1)

In [68]:
# a and b are tiled, line by line for a, column by column for b, to enable the add operation
c = a+b
c

array([[ 6,  7,  8,  9],
       [ 7,  8,  9, 10],
       [ 8,  9, 10, 11]])

In [69]:
c.shape

(3, 4)

## Linear Algebra
### 1D arrays

* `numpy.inner(a, b)`
Return the inner/scalar product of 2 vectors
* `numpy.outer(a, b)`
Return the outer product of 2 vectors

In [70]:
a = np.array([1, 2, 3, 4])
a

array([1, 2, 3, 4])

In [71]:
b = np.array([5, 6, 7, 8])
b

array([5, 6, 7, 8])

In [72]:
np.inner(a, b)

70

In [73]:
np.outer(a, b)

array([[ 5,  6,  7,  8],
       [10, 12, 14, 16],
       [15, 18, 21, 24],
       [20, 24, 28, 32]])

### 2D arrays

* `a.T` is the transposition of `a`
* `numpy.dot(a, b)` return the matrix product between a and b.

In [74]:
a = np.random.randn(3, 5)
a.T

array([[ 0.2576092 ,  0.47543723, -0.82207224],
       [-0.07447876, -1.08697823, -0.0768496 ],
       [-0.00607162,  0.56257805,  0.0792357 ],
       [-1.51167531, -0.60503362,  0.9520792 ],
       [ 0.0949625 ,  0.29937258, -0.64353791]])

In [75]:
a = np.random.randn(3, 5)
b = np.random.randn(5, 2)
np.dot(a, b)

array([[-0.57993687,  0.53396444],
       [ 0.28427718, -3.2148884 ],
       [-1.08575208,  3.4177859 ]])

In [76]:
# Equivalent notation
a.dot(b)

array([[-0.57993687,  0.53396444],
       [ 0.28427718, -3.2148884 ],
       [-1.08575208,  3.4177859 ]])

In [77]:
# Since python 3.5, the @ symbol can be used for matrix multiplication
a @ b

array([[-0.57993687,  0.53396444],
       [ 0.28427718, -3.2148884 ],
       [-1.08575208,  3.4177859 ]])

## Saving and loading data

### Input

* `numpy.load(file, mmap_mode=None, allow_pickle=True, fix_imports=True, encoding='ASCII')`

load a `npy` or `npz` file,
* `numpy.loadtxt(fname, dtype=<type 'float'>, comments='#', delimiter=None, converters=None, skiprows=0, usecols=None, unpack=False, ndmin=0)`

load a `txt` file.

### Output

* `numpy.save(file, arr, allow_pickle=True, fix_imports=True)`

save ONE array into a `npy` file,

* `numpy.savez(file, *args, **kwds)`

save many arrays into an `npz` file,

* `numpy.savetxt(fname, X, fmt='%.18e', delimiter=' ', newline='\n', header='', footer='', comments='# ')`

save ONE array into a `txt` file,


## Your turn

Try to answer each following questions by a small snippet of code.

1. How to reverse a vector (1d array) ?

2. How to keep dimension consistency when slicing a matrix (2d array) ?

3. How to create a (5,5) array with random values and find the extrema values ?

4. With the help of broadcasting, how to produce a matrix A where A\[i,j\] = 2i + j ? (no for loop allowed)

5.  A is a (4,4) int array, I want to change the last element of A to 1.5 without loosing any information. How can I do it ?